### ISSUE TITLE
The number of samples in the dataset

### ISSUE THREAD

**Issue by zhangmuzhi666** [alirezaabdollahpour/SuperDeepFool#3]
Regarding the details in your open-source GitHub code, I would like to ask:

1. For MNIST, CIFAR10, and ImageNet datasets used in your experiments, Was this done specifically to ensure 100% classifier accuracy as the basis for the experiments?
2. When I tried to repeat the experiments with 100, 1000, and 10,000 samples, the classifier's accuracy dropped a lot. At the same time, the success rates of both SuperDeepFool and DeepFool attacks also decreased. This doesn't match the 100% accuracy shown in Table 4 of your paper.

Please explain this difference because it's important for other people to repeat your research. If there are any settings or parameters I missed, please let me know.

**Comment by alirezaabdollahpour** [comment]
# Model Evaluation Protocol

Thank you for your inquiry regarding our evaluation methodology.

## Clean Accuracy Verification
We first validate that each model's clean accuracy aligns with the reported benchmarks in their respective original publications. This establishes a baseline for subsequent robustness analysis.

## Robust Accuracy Assessment
For robust accuracy evaluation, we implement a selective testing protocol wherein we exclude any samples misclassified by the models under standard conditions. This methodological constraint ensures that our adversarial perturbations target only correctly classified samples.

## Implementation Notes
The provided code~(main.py) represents a simplified algorithmic framework for preliminary experimentation on baseline models. For comprehensive evaluation across diverse architectural paradigms, we recommend developing a custom implementation that rigorously maintains consistency between:

1. The normalization parameters used during evaluation
2. The specific normalization schema employed during model training

Once this alignment is established, you can deploy the SuperDeepFool Attack module to execute the attack procedure exclusively on the correctly classified subset of samples.

**Comment by zhangmuzhi666** [comment]
Thank you for your detailed reply earlier. I’ve understood that the current experiment is conducted based on a sample set with misclassified samples excluded. However, regarding the experiment’s rigor, I still have questions: Could this sample processing approach potentially lack rigor?
Additionally, I haven’t fully grasped the normalization-related content. In the code you provided, when using load_cifar10 with the default PREPROCESSINGS[None], this operation appears to omit normalization. After normalizing the Standard model, its accuracy on the complete dataset drops to 28.25%. Yet, with the unnormalized basic code you shared, the accuracy reaches 94.77%. This notable discrepancy has left me quite puzzled.
Would it be possible for you to kindly spare some time to clarify these concerns?

**Comment by alirezaabdollahpour** [comment]
Hi,

Regarding the normalization process, it depends on the model's properties you're using. 

For instance, if you use RobustBench models, you must verify the normalization specifications for those particular models. Similarly, if you use your own trained models, you must apply the specific normalization parameters used during the training process. It's essential to always verify normalization correctness by checking clean accuracy before executing any attack. 

Let's assume you overlooked clean accuracy verification and directly attacked the model. Two crucial factors can help validate the correctness of your experiments:

1- The $\ell_{2}$ norms of perturbations.
   If you examine the $\ell_{2}$ norms of perturbations for CIFAR-10, they should consistently fall near 0.1 and 0.2 for non-adversarially trained networks in the optimal case. What does this imply? It indicates that near-optimal $\ell_{2}$ norm perturbations across different optimal attacks tend to be similar. This consistency stems from the fact that these methods are all non-convex optimization algorithms known to converge to near-optimal solutions. Therefore, the perturbation norms in the optimal case should not deviate significantly across algorithms like Adaptive-PGD, FMN, FAB, DDN, and SuperDeepFool. It's important to note that this pattern holds specifically for small-scale datasets like CIFAR-10 and CIFAR-100 when using non-adversarially trained networks. For adversarially-trained networks, the average $\ell_{2}$ perturbation will exhibit higher variance across different attack methods. As a first sanity check, you should always assess the reasonableness of your attack's $\ell_{2}$ perturbation norms.

2- The number of iterations per sample for non-fixed iteration attacks.
   We know that fixed-iteration attacks like Adaptive-PGD, FMN, FAB, and DDN typically use 100 iterations per sample. This predefined count provides little diagnostic information about attack correctness, so you must rely on point 1 mentioned above. However, you can analyze iteration counts for adaptive methods like DeepFool or SuperDeepFool. If DeepFool converges (finds an adversarial example) in just 2-3 iterations, it likely indicates that the sample was already misclassified by the model. A critical requirement here is to set the number of line searches for the algorithm to zero to ensure you're accurately profiling iteration counts for DeepFool or SuperDeepFool.

How to handle the normalization process automatically in large-scale experiments?
For a solution, always check this ["link"](https://github.com/RobustBench/robustbench/blob/master/robustbench/model_zoo/cifar10.py) for correct normalization protocols.

What kinds of discrepancies might you encounter with the RobustBench library?
You should always verify the version of the RobustBench library you're using.
In some versions, the model downloading process has issues that can lead to corrupted checkpoints. You can resolve this by consulting this [link](https://github.com/RobustBench/robustbench/issues/177).


If you have any other questions, please don't hesitate to contact me.

**Comment by zhangmuzhi666** [comment]
Dear Dr. Alireza Abdollahpour,

I am writing to express my deepest gratitude for your expert guidance on standardization protocols in machine learning workflows. Your profound insights have been instrumental in resolving critical experimental validation challenges.

During the evaluation of the standard classifier model from your code repository, I observed a striking phenomenon: ​the model achieved 94.77% classification accuracy without normalization, which plummeted to 28.25% when standardized normalization was applied. This drastic discrepancy strongly indicates a systematic misalignment between preprocessing parameters used during training and those implemented during inference. Specifically, the model architecture appears to lack integrated normalization layers or synchronized dataset-specific statistical parameters (e.g., CIFAR-10's μ=[0.4914, 0.4822, 0.4465], σ=[0.2470, 0.2435, 0.2616]) in its original training configuration. This finding conclusively validates your core principle that "normalization integrity must be protocol-driven rather than assumption-based."

Your rigorous methodological framework has not only resolved this technical anomaly but also established a reproducible validation framework for my ongoing research. Should any implementation ambiguities emerge in future studies, I shall highly value your expert counsel.

Thank you once again for your exceptional scholarly mentorship and timely assistance

**Comment by alirezaabdollahpour** [comment]
Hi @color6789,

I really appreciate those warm words of yours! 

But whoops - "Doctor" isn't on my business card yet :)) I'm just a 23-year-old human-in-progress who (if I will be alive) should be rocking that fancy "Dr." title about 5 years from now! 🎓⏳

If you have any other concerns, do not hesitate to ping me.

**Comment by zhangmuzhi666** [comment]
Haha, your kind words mean the world✨! May your PhD journey be smooth and fulfilling until you proudly rock that "Dr." title🚀~ Can't wait for more inspiring conversations with you in the years to come! Let's definitely stay connected😊

**Comment by zhangmuzhi666** [comment]
.

**Comment by zhangmuzhi666** [comment]
My friend Alireza Abdollahpour, there's a problem that has puzzled me for a long time. In your DeepFool code, the iteration always stops before the  steps (step), and according to the code logic, all samples should be successfully attacked at this point. When I predict each adversarial example in the generated list one by one, the success rate indeed reaches 100%; however, when these examples are input into the model in a batch (batch size), the success rate drops below 100%. What is the reason behind this contradictory phenomenon? I hope you can help me understand, my friend. 

1. I have a hunch that the conversion from a list to a tensor could potentially lead to information loss. Additionally, it's possible that the problem lies in the batch size setting of the model.




My code as：  

```Python
def forward(self, images, labels, return_target_labels=False):
        r"""
        Overridden.
        """
        self.model.eval()
        images = images.clone().detach().to(self.device)
        labels = labels.clone().detach().to(self.device)

        batch_size = len(images)
        # print("batch_size:", batch_size)
        correct = torch.tensor([True] * batch_size)
        target_labels = labels.clone().detach().to(self.device)
        curr_steps = 0

        adv_images = []
        for idx in range(batch_size):
            image = images[idx : idx + 1].clone().detach()
            adv_images.append(image)
        while (True in correct) and (curr_steps < self.steps):
            for idx in range(batch_size):
                if not correct[idx]:
                    continue
                early_stop, pre, adv_image = self._forward_indiv(
                    adv_images[idx], labels[idx]
                )
                adv_images[idx] = adv_image
                target_labels[idx] = pre
                if early_stop:
                    correct[idx] = False
            curr_steps += 1

        self.number_of_iterations = curr_steps
        right1 = 0
        for i in range(batch_size):
            if target_labels[i] != labels[i]:
                right1 += 1
        print("right1:", right1)
        right2 = 0
        for t in range(batch_size):
            pred = self.model(adv_images[t]).argmax(dim=1)
            if pred != labels[t]:
                right2 += 1
        print("right2:", right2)
        
        adv_images = torch.cat(adv_images).detach()
        preds = self.model(adv_images).argmax(dim=1)
        right3 = (preds != labels).sum().item()
        print("right3:", right3)
        right4=0
        for j  in range (batch_size):
            pred=self.model(adv_images[j:j+1]).argmax(dim=1)
            if pred != labels[j]:
               right4+=1
        print("right4:", right4)
                
        if return_target_labels:

            return adv_images, target_labels

        return adv_images, curr_steps
```

**Comment by zhangmuzhi666** [comment]
Figure 3 

Dear friend, when validating the overstepping of the DF perturbation in your paper, why does it state in your paper that when γ is close to 1, the fooling rate will drop sharply? Shouldn't it be increasing? You also made the same statement in the subsequent experimental description section. Could you please provide me with a response regarding this?
